### ISSUE TITLE
Boost python exceptions crashing engines

### ISSUE THREAD

**Issue by bluescarni** [ipython/ipyparallel#236]
I am trying to use ipyparallel to parallelize the workload in a Python project largely consisting of a C++ codebase exposed to Python via Boost.Python. I am using a local ipcluster for testing:

```console
ipcluster start -n 4 --debug
```

I am using the task based interface:

```python
>>> import ipyparallel
>>> rc = ipyparallel.Client()
>>> lview = rc.load_balanced_view()
>>> ar = lview.apply(...)
>>> ar.get()
```

Everything seems to work properly in my initial testing, until I try to distribute a task in which C++ code throws an exception (FYI, Boost.Python has a mechanism in which C++ exceptions are caught at the boundary between C++ and Python, and re-thrown in Python as Python exceptions -- in many years of using Boost Python, this mechanism has always been working flawlessly for my use cases).

When I try to distribute a task that throws from C++, I can see the following happening in the debug output of ipcluster:

```console
2017-03-05 22:07:17.940 [IPClusterStart] Process '/usr/bin/python3.5' stopped: {'pid': 18580, 'exit_code': -6}
2017-03-05 22:07:21.520 [IPClusterStart] b"2017-03-05 22:07:21.520 [IPControllerApp] heartbeat::missed b'af9a4de7-e140-4a91-b4bb-c7089b04adc9' : 1"
2017-03-05 22:07:24.520 [IPClusterStart] b"2017-03-05 22:07:24.520 [IPControllerApp] heartbeat::missed b'af9a4de7-e140-4a91-b4bb-c7089b04adc9' : 2"
2017-03-05 22:07:27.520 [IPClusterStart] b"2017-03-05 22:07:27.520 [IPControllerApp] heartbeat::missed b'af9a4de7-e140-4a91-b4bb-c7089b04adc9' : 3"
2017-03-05 22:07:30.520 [IPClusterStart] b"2017-03-05 22:07:30.520 [IPControllerApp] heartbeat::missed b'af9a4de7-e140-4a91-b4bb-c7089b04adc9' : 4"
2017-03-05 22:07:33.520 [IPClusterStart] b"2017-03-05 22:07:33.520 [IPControllerApp] heartbeat::missed b'af9a4de7-e140-4a91-b4bb-c7089b04adc9' : 5"
2017-03-05 22:07:36.520 [IPClusterStart] b"2017-03-05 22:07:36.520 [IPControllerApp] heartbeat::missed b'af9a4de7-e140-4a91-b4bb-c7089b04adc9' : 6"
2017-03-05 22:07:39.520 [IPClusterStart] b"2017-03-05 22:07:39.520 [IPControllerApp] heartbeat::missed b'af9a4de7-e140-4a91-b4bb-c7089b04adc9' : 7"
2017-03-05 22:07:42.520 [IPClusterStart] b"2017-03-05 22:07:42.520 [IPControllerApp] heartbeat::missed b'af9a4de7-e140-4a91-b4bb-c7089b04adc9' : 8"
2017-03-05 22:07:45.520 [IPClusterStart] b"2017-03-05 22:07:45.520 [IPControllerApp] heartbeat::missed b'af9a4de7-e140-4a91-b4bb-c7089b04adc9' : 9"
2017-03-05 22:07:48.520 [IPClusterStart] b"2017-03-05 22:07:48.520 [IPControllerApp] heartbeat::missed b'af9a4de7-e140-4a91-b4bb-c7089b04adc9' : 10"
2017-03-05 22:07:51.520 [IPClusterStart] b"2017-03-05 22:07:51.520 [IPControllerApp] heartbeat::missed b'af9a4de7-e140-4a91-b4bb-c7089b04adc9' : 11"
2017-03-05 22:07:51.521 [IPClusterStart] b'2017-03-05 22:07:51.520 [IPControllerApp] registration::unregister_engine(3)'
2017-03-05 22:07:56.531 [IPClusterStart] b'ERROR:tornado.application:Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x7f3ce0377598>)'
2017-03-05 22:07:56.535 [IPClusterStart] b'Traceback (most recent call last):'
2017-03-05 22:07:56.535 [IPClusterStart] b'  File "/usr/lib64/python3.5/site-packages/tornado/ioloop.py", line 604, in _run_callback'
2017-03-05 22:07:56.535 [IPClusterStart] b'    ret = callback()'
2017-03-05 22:07:56.536 [IPClusterStart] b'  File "/usr/lib64/python3.5/site-packages/tornado/stack_context.py", line 275, in null_wrapper'
2017-03-05 22:07:56.536 [IPClusterStart] b'    return fn(*args, **kwargs)'
2017-03-05 22:07:56.536 [IPClusterStart] b'  File "/usr/lib64/python3.5/site-packages/ipyparallel/controller/scheduler.py", line 325, in <lambda>'
2017-03-05 22:07:56.536 [IPClusterStart] b'    lambda : self.handle_stranded_tasks(uid),'
2017-03-05 22:07:56.536 [IPClusterStart] b'  File "/usr/lib64/python3.5/site-packages/ipyparallel/controller/scheduler.py", line 335, in handle_stranded_tasks'
2017-03-05 22:07:56.536 [IPClusterStart] b'    for msg_id in lost.keys():'
2017-03-05 22:07:56.537 [IPClusterStart] b'RuntimeError: dictionary changed size during iteration'
2017-03-05 22:07:56.537 [IPClusterStart] b"2017-03-05 22:07:56.534 [IPControllerApp] task::task '61317ff8-c759-4faa-aebb-2fd57595c586' finished on 3"
```

On the ipython side, when I try to ``.get()`` the result from the future, the interpreter waits for a bit and then says:

```console
In [14]: ar.get()
Traceback (most recent call last):
  File "/usr/lib64/python3.5/site-packages/ipyparallel/controller/scheduler.py", line 347, in handle_stranded_tasks
    raise error.EngineError("Engine %r died while running task %r"%(engine, msg_id))
ipyparallel.error.EngineError: Engine b'af9a4de7-e140-4a91-b4bb-c7089b04adc9' died while running task '61317ff8-c759-4faa-aebb-2fd57595c586'
```

I am able to reproduce the problem with a minimal Boost.Python module consisting of a single class with a single method which throws an ``std::invalid`` exception, which normally is translated into a ``ValueError`` python exception. I can make this minimal example available in a github repo if it can help to debug the issue. The exception translation works fine from the ipython prompt:

```console
In [1]: import ipy_testmod

In [2]: s = ipy_testmod.my_struct()

In [3]: s.my_method(123)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-3-ba5412712f47> in <module>()
----> 1 s.my_method(123)

ValueError: error!

In [4]:
```

It just breaks down when thrown from an ipyparallel engine:

```console
In [7]: def func(x, n):
   ...:     return x.my_method(n)
   ...: 

In [8]: ar = lview.apply(func, s, 123)

In [9]: ar.get()
Traceback (most recent call last):
  File "/usr/lib64/python3.5/site-packages/ipyparallel/controller/scheduler.py", line 347, in handle_stranded_tasks
    raise error.EngineError("Engine %r died while running task %r"%(engine, msg_id))
ipyparallel.error.EngineError: Engine b'03245294-8978-4b6c-a3c6-abf4e0cb7710' died while running task '936c8b82-bbf9-4723-955b-657c048d93e3'
```

Wrapping the throwing code into a ``try/except`` block does not seem to help:

```console
In [10]: def func(x, n):
    ...:     try:
    ...:         return x.my_method(n)
    ...:     except:
    ...:         return 23
    ...: 

In [11]: ar = lview.apply(func, s, 123)

In [12]: ar.get()
Traceback (most recent call last):
  File "/usr/lib64/python3.5/site-packages/ipyparallel/controller/scheduler.py", line 347, in handle_stranded_tasks
    raise error.EngineError("Engine %r died while running task %r"%(engine, msg_id))
ipyparallel.error.EngineError: Engine b'91e8abe8-2371-469c-af4b-f6d376ed3e5e' died while running task '3b1e0902-a321-4d06-affa-d9417c18b2d4'

In [13]:
```

It seems like just the act of throwing from C++ brings down the engine.

**Comment by minrk** [comment]
I don't know about the objects you are using, but the only thing I see is that the object is pickled and sent to the engine before calling `.my_method()`. Do you see the same error if you create the function and call the failing method entirely on the engine (i.e. never pass the object itself to apply)?

I don't know how boost's exception wrapping works, but I could imagine either some local initialization isn't getting triggered in the remote environment and/or when the object is serialized / deserialized with pickle, it somehow gets detached from whatever does this.

**Comment by bluescarni** [comment]
@minrk Thanks for the reply!

How would I create the object and run the offending function directly on the engine? Can I somehow connect to the engines after they have been started with ipcluster?

**Comment by bluescarni** [comment]
@minrk just a wild guess, but is it possible that ipyparallel or some of its dependencies have been compiled with ``-fno-exceptions`` or similar flags?

**Comment by bluescarni** [comment]
@minrk I just converted my minimal C++ module to use pybind11 (https://github.com/pybind/pybind11) instead of Boost.Python for the exposition. I have exactly the same behaviour: exception conversion from C++ to Python works as expected in the ipython prompt, from a script, etc., but it crashes the ipyparallel engine in exactly the same way as Boost.Python does.

This seems to be more of a fundamental problem with C++ exceptions rather than specifically with Boost.Python or pybind11.

**Comment by minrk** [comment]
> How would I create the object and run the offending function directly on the engine?

something like:

```python
def f():
    obj = instantiate_my_object()
    return obj.my_method(value)

view.apply(f)
```

>  Can I somehow connect to the engines after they have been started with ipcluster?

Yes, you can use the ipyparallel Client to do this. You can send functions to be evaluated on the engines. The main thing I'm trying to test is whether the object being serialized is relevant. So if you can write a function that takes no arguments and reproduces the issue where it crashes remotely and runs locally, that would be the most useful.

> is it possible that ipyparallel or some of its dependencies have been compiled with -fno-exceptions or similar flags?

I don't think so. IPython Parallel isn't compiled at all (it's Python), and there aren't any dependencies on the engines that aren't also present in the client, so if you can call it locally just before submitting it with apply, I don't think that would be it.

Are you running a local cluster (engines and client on the same machine), or are the engines remote?

**Comment by bluescarni** [comment]
@minrk Skipping the serialization bit still results in the engine crashing:

```console
In [1]: import ipyparallel

In [2]: rc = ipyparallel.Client()

In [3]: lview = rc.load_balanced_view()

In [4]: def func():
   ...:     import ipy_testmod
   ...:     ipy_testmod.my_struct().my_method(123)
   ...:     

In [5]: ar = lview.apply(func)

In [6]: ar.get()
Traceback (most recent call last):
  File "/usr/lib64/python3.5/site-packages/ipyparallel/controller/scheduler.py", line 347, in handle_stranded_tasks
    raise error.EngineError("Engine %r died while running task %r"%(engine, msg_id))
ipyparallel.error.EngineError: Engine b'f7d939fe-8260-4888-b5e7-0398d512617d' died while running task 'b9dfa4ce-2c41-4e56-a08e-56da13a54875'

In [7]: func() # NOTE: this works as expected, translates the C++ exception into ValueError
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-7-08a2da4138f6> in <module>()
----> 1 func()

<ipython-input-4-985d7752d976> in func()
      1 def func():
      2     import ipy_testmod
----> 3     ipy_testmod.my_struct().my_method(123)
      4 

ValueError: error!

In [8]:
```

The output of ipcluster:

```console
2017-03-06 17:10:38.326 [IPClusterStart] b'2017-03-06 17:10:38.326 [IPControllerApp] registration::finished registering engine 0:6763df59-10aa-456d-88ed-22a18a5c71c4'
2017-03-06 17:10:38.328 [IPClusterStart] b'2017-03-06 17:10:38.327 [IPControllerApp] engine::Engine Connected: 0'
2017-03-06 17:10:38.328 [IPClusterStart] b'2017-03-06 17:10:38.328 [IPControllerApp] registration::finished registering engine 3:30661364-375c-4827-92a8-cf0abb6409eb'
2017-03-06 17:10:38.329 [IPClusterStart] b'2017-03-06 17:10:38.328 [IPControllerApp] engine::Engine Connected: 3'
2017-03-06 17:10:38.329 [IPClusterStart] b'2017-03-06 17:10:38.329 [IPControllerApp] registration::finished registering engine 2:71319e88-8f8e-47f7-94fc-cdf2979ef70f'
2017-03-06 17:10:38.330 [IPClusterStart] b'2017-03-06 17:10:38.330 [IPControllerApp] engine::Engine Connected: 2'
2017-03-06 17:10:38.330 [IPClusterStart] b'2017-03-06 17:10:38.330 [IPControllerApp] registration::finished registering engine 1:f7d939fe-8260-4888-b5e7-0398d512617d'
2017-03-06 17:10:38.331 [IPClusterStart] b'2017-03-06 17:10:38.331 [IPControllerApp] engine::Engine Connected: 1'
2017-03-06 17:10:42.966 [IPClusterStart] b"2017-03-06 17:10:42.965 [IPControllerApp] client::client b'\\x00k\\x8bEh' requested 'connection_request'"
2017-03-06 17:10:42.966 [IPClusterStart] b"2017-03-06 17:10:42.965 [IPControllerApp] client::client [b'\\x00k\\x8bEh'] connected"
2017-03-06 17:11:02.652 [IPClusterStart] Engines appear to have started successfully
2017-03-06 17:11:08.010 [IPClusterStart] b"2017-03-06 17:11:08.009 [IPControllerApp] task::task 'b9dfa4ce-2c41-4e56-a08e-56da13a54875' arrived on 1"
2017-03-06 17:11:08.106 [IPClusterStart] Process '/usr/bin/python3.5' stopped: {'exit_code': -6, 'pid': 26654}
2017-03-06 17:11:11.326 [IPClusterStart] b"2017-03-06 17:11:11.326 [IPControllerApp] heartbeat::missed b'f7d939fe-8260-4888-b5e7-0398d512617d' : 1"
2017-03-06 17:11:14.326 [IPClusterStart] b"2017-03-06 17:11:14.326 [IPControllerApp] heartbeat::missed b'f7d939fe-8260-4888-b5e7-0398d512617d' : 2"
2017-03-06 17:11:17.327 [IPClusterStart] b"2017-03-06 17:11:17.326 [IPControllerApp] heartbeat::missed b'f7d939fe-8260-4888-b5e7-0398d512617d' : 3"
2017-03-06 17:11:20.326 [IPClusterStart] b"2017-03-06 17:11:20.326 [IPControllerApp] heartbeat::missed b'f7d939fe-8260-4888-b5e7-0398d512617d' : 4"
2017-03-06 17:11:23.326 [IPClusterStart] b"2017-03-06 17:11:23.326 [IPControllerApp] heartbeat::missed b'f7d939fe-8260-4888-b5e7-0398d512617d' : 5"
2017-03-06 17:11:26.326 [IPClusterStart] b"2017-03-06 17:11:26.326 [IPControllerApp] heartbeat::missed b'f7d939fe-8260-4888-b5e7-0398d512617d' : 6"
2017-03-06 17:11:29.327 [IPClusterStart] b"2017-03-06 17:11:29.326 [IPControllerApp] heartbeat::missed b'f7d939fe-8260-4888-b5e7-0398d512617d' : 7"
2017-03-06 17:11:32.326 [IPClusterStart] b"2017-03-06 17:11:32.326 [IPControllerApp] heartbeat::missed b'f7d939fe-8260-4888-b5e7-0398d512617d' : 8"
2017-03-06 17:11:35.327 [IPClusterStart] b"2017-03-06 17:11:35.326 [IPControllerApp] heartbeat::missed b'f7d939fe-8260-4888-b5e7-0398d512617d' : 9"
2017-03-06 17:11:38.326 [IPClusterStart] b"2017-03-06 17:11:38.326 [IPControllerApp] heartbeat::missed b'f7d939fe-8260-4888-b5e7-0398d512617d' : 10"
2017-03-06 17:11:41.327 [IPClusterStart] b"2017-03-06 17:11:41.326 [IPControllerApp] heartbeat::missed b'f7d939fe-8260-4888-b5e7-0398d512617d' : 11"
2017-03-06 17:11:41.327 [IPClusterStart] b'2017-03-06 17:11:41.327 [IPControllerApp] registration::unregister_engine(1)'
2017-03-06 17:11:46.339 [IPClusterStart] b'ERROR:tornado.application:Exception in callback functools.partial(<function wrap.<locals>.null_wrapper at 0x7fcc40cae730>)'
2017-03-06 17:11:46.341 [IPClusterStart] b'Traceback (most recent call last):'
2017-03-06 17:11:46.342 [IPClusterStart] b'  File "/usr/lib64/python3.5/site-packages/tornado/ioloop.py", line 604, in _run_callback'
2017-03-06 17:11:46.342 [IPClusterStart] b'    ret = callback()'
2017-03-06 17:11:46.342 [IPClusterStart] b'  File "/usr/lib64/python3.5/site-packages/tornado/stack_context.py", line 275, in null_wrapper'
2017-03-06 17:11:46.342 [IPClusterStart] b'    return fn(*args, **kwargs)'
2017-03-06 17:11:46.342 [IPClusterStart] b'  File "/usr/lib64/python3.5/site-packages/ipyparallel/controller/scheduler.py", line 325, in <lambda>'
2017-03-06 17:11:46.342 [IPClusterStart] b'    lambda : self.handle_stranded_tasks(uid),'
2017-03-06 17:11:46.342 [IPClusterStart] b'  File "/usr/lib64/python3.5/site-packages/ipyparallel/controller/scheduler.py", line 335, in handle_stranded_tasks'
2017-03-06 17:11:46.343 [IPClusterStart] b'    for msg_id in lost.keys():'
2017-03-06 17:11:46.343 [IPClusterStart] b'RuntimeError: dictionary changed size during iteration'
2017-03-06 17:11:46.343 [IPClusterStart] b"2017-03-06 17:11:46.341 [IPControllerApp] task::task 'b9dfa4ce-2c41-4e56-a08e-56da13a54875' finished on 1"
```

(I have included some extra debug output before the failure, just in case)

I am running a local cluster, same machine:

```console
ipcluster start -n 4 --debug
```

**Comment by bluescarni** [comment]
And here is a gist with my minimal pybind11 module:

https://gist.github.com/bluescarni/49cea7bc085e72863ebd47b91344ba0a

**Comment by bluescarni** [comment]
@minrk How would I hook up GDB to the ipyparallel engines? I'd like to try to see if I can get any additional clues about what is going on, but I rarely use GDB and all my attempts so far have not been fruitful.

**Comment by bluescarni** [comment]
@minrk This is interesting: if I launch ``ipcluster`` like this:

```console
LD_PRELOAD=/usr/lib/gcc/x86_64-pc-linux-gnu/6.3.0/libstdc++.so.6 ipcluster start -n 4 --debug
```

it seems like everything works as intended. So it seems like something is somehow wrong with the C++ runtime inside the engine?

**Comment by minrk** [comment]
If you skip ipcluster, you can run with gdb:

```bash
[shell 1] $ ipcontroller
[shell 2] $ gdb --args ipengine
```

Even without gdb, running the engine directly might be informative.

Are your engines on the same machine running in the same env as your client?

**Comment by bluescarni** [comment]
@minrk yes everything happens on the same machine in the same environment. The packages are installed via the distribution's package manager.

I just tried my minimal module on another linux system, an Arch Linux installation, and I did not have any issue. At this point I am positive that the problem is specific to my Gentoo system, probably something to do either with installation paths or maybe due to the fact that I am using the gold linker instead of the standard binutils linker.

If you don't need me to investigate further, I can close the ticket.
